src/lib/storage/disk_analyzer.rb

```diff
@@ -0,0 +1,368 @@
+      DEFAULT_DISK_CHECK_LIMIT = 10
+
+      attr_reader :installation_disks, :candidate_disks
+      attr_reader :windows_partitions, :linux_partitions
+      attr_accessor :disk_check_limit
+
+      def initialize
+        Yast.import "Arch"
+
+        @installation_disks = []
+        @candidate_disks    = []
+        @linux_partitions   = []
+        @windows_partitions = [] # only filled if @linux_partitions is empty
+
+        # Maximum number of disks to check. This might be important on
+        # architectures that tend to have very many disks (s390).
+        @disk_check_limit = DEFAULT_DISK_CHECK_LIMIT
+      end
+
+      # Analyze disks and partitions. Make sure to call this before querying
+      # any member variables.
+      #
+      def analyze
+        @installation_disks = find_installation_disks
+        @candidate_disks    = find_candidate_disks
+        @linux_partitions   = find_linux_partitions
+
+        if @linux_partitions.empty?
+          @windows_partitions = find_windows_partitions
+        else
+          # We only want to resize any MS Windows partition if there is no
+          # Linux already on that machine, so we don't need to check for
+          # Windows partitions in that case. This saves mounting and unmounting
+          # all Windows-like partitions.
+          log.info("Linux partitions found - not checking for Windows partitions")
+        end
+
+        log.info("Installation disks: #{dev_names(@installation_disks)}")
+        log.info("Candidate    disks: #{dev_names(@candidate_disks)}")
+        log.info("Linux   partitions: #{dev_names(@linux_partitions)}")
+        log.info("Windows partitions: #{dev_names(@windows_partitions)}")
+      end
+
+      # Find disks that look like the current installation medium
+      # (the medium we just booted from to start the installation).
+      #
+      # This is limited with DiskAnalyzer::disk_check_limit because some
+      # architectures (s/390) tend to have a large number of disks, and
+      # checking if a disk is an installation disk involves mounting and
+      # unmounting each partition on that disk.
+      #
+      # @return [Array<::Storage::Disk>] installation disks
+      #
+      def find_installation_disks
+        disks = StorageManager.instance.probed.all_disks.to_a
+        # FIXME: to_a should not be necessary:
+        # libstorage should return something that Ruby can handle.
+        # This is very likely a problem of the Swig bindings.
+
+        usb_disks, non_usb_disks = disks.partition { |disk| disk.transport == ::Storage::USB }
+        disks = usb_disks + non_usb_disks
+        disk_count = 0
+        installation_disks = []
+
+        disks.each do |disk|
+          installation_disks << disk if installation_disk?(disk)
+          disk_count += 1
+          if disk_count > @disk_check_limit
+            log.info("Disk check limit reached after #{disk}")
+            break
+          end
+        end
+        installation_disks
```

Comment:
Yikes. That's what people hate about Ruby. Why change perfectly well readable code into gibberish without a pressing need to do so?

It begins with the sorting: I could never tell which stuff gets sorted first. The old storage proposal code is full of things like that; you'd have to tear it apart, add logging and execute it to find out what the hell this does. OTOH that

    usb_disks, non_usb_disks = disks.partition { |disk| disk.transport == ::Storage::USB }
    disks = usb_disks + non_usb_disks

is self-explanatory IMHO.

Also, that changed limit check and its logging will not tell me what disks were considered and which ones weren't - since it doesn't output even the device names. But then, you are right, that output is missing in my version as well. I'll change that.